### PR TITLE
just match for BinaryExpr Date/Time +/- Interval

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #.

# Rationale for this change

- remove result in temporal_add_sub_coercion()
- just match for BinaryExpr Date/Time +/- Interval

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -217,57 +217,61 @@ pub fn temporal_add_sub_coercion(
...
         // if two date/timestamp are being added/subtracted, return an error indicating that the operation is not supported
-        (lhs, rhs, _) if (is_date(lhs) || is_timestamp(lhs)) && (is_date(rhs) || is_timestamp(rhs)) => {
-            Err(DataFusionError::Plan(format!(
-                "{lhs_type:?} {rhs_type:?} is an unsupported operation. addition/subtraction on dates/timestamps only supported with interval types"
-            )))
+        (lhs, rhs, Operator::Minus)
+            if (is_date(lhs) || is_timestamp(lhs))
+                && (is_date(rhs) || is_timestamp(rhs)) =>
+        {
+            // TODO: support Minus
```
Comment (comphead) [inline datafusion/expr/src/type_coercion/binary.rs:242]: This None might be misleading if Minus called? Probably we can return `Err` on the `temporal_add_sub_coercion` caller side if Minus called?

Comment (jackwener) [reply datafusion/expr/src/type_coercion/binary.rs:242]: None will be convert to Err outside.
Here use None is right. If here return Err, other None also need to be Err

Comment (comphead) [reply datafusion/expr/src/type_coercion/binary.rs:242]: in that case the comment doesn't make sense 
https://github.com/apache/arrow-datafusion/pull/5932/files#diff-6a1eb5571196f1bd9f3defbbd97ed99e0e527e4ede426ce7e802ab949f1c431eR237

Comment (jackwener) [reply datafusion/expr/src/type_coercion/binary.rs:242]: Yes, these code exist problem, I will fix it in other PR.
In fact, we should allow datetime - datetime

Comment (waynexia) [reply datafusion/expr/src/type_coercion/binary.rs:242]: And this comment https://github.com/apache/arrow-datafusion/pull/5932/files#diff-6a1eb5571196f1bd9f3defbbd97ed99e0e527e4ede426ce7e802ab949f1c431eR214-R215

**Comment by alamb** [comment]
cc @berkaysynnada  who I think added the coercion for timestamp related functions

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 2]
```diff
@@ -242,20 +242,19 @@ impl TreeNodeRewriter for TypeCoercionRewriter {
...
-                    | (
+                    ) if matches!(op, Operator::Plus | Operator::Minus) => Ok(expr),
+                    (
                         &DataType::Interval(_),
                         DataType::Date32 | DataType::Date64 | DataType::Timestamp(_, _),
-                    ) => {
-                        // this is a workaround for https://github.com/apache/arrow-datafusion/issues/3419
-                        Ok(expr.clone())
-                    }
+                    ) if matches!(op, Operator::Plus) => Ok(expr),
```
Comment (waynexia) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:257]: 👍 good catch

File: datafusion/expr/src/type_coercion/binary.rs [hunk 3]
```diff
@@ -276,18 +280,16 @@ fn handle_timestamp_minus(lhs: &DataType, rhs: &DataType) -> Result<Option<DataT
...
         )
         | (
             DataType::Timestamp(TimeUnit::Nanosecond, _),
             DataType::Timestamp(TimeUnit::Nanosecond, _),
-        ) => Ok(Some(DataType::Interval(IntervalUnit::MonthDayNano))),
-        (_, _) => Err(DataFusionError::Plan(
-            "The timestamps have different types".to_string(),
-        )),
+        ) => Some(DataType::Interval(IntervalUnit::MonthDayNano)),
+        (_, _) => None,
```
Comment (waynexia) [inline datafusion/expr/src/type_coercion/binary.rs:292]: To my understanding this is only "not supported yet", right? Or did I miss something as the function comment says this is an invalid case

**Review by waynexia — APPROVED** [review]
Looking good 👍

File: datafusion/expr/src/type_coercion/binary.rs (hunk 3 above)
Comment (berkaysynnada) [reply datafusion/expr/src/type_coercion/binary.rs:292]: Yes, the subtraction of different timestamp types needs special handling. I think if the operation is done with second or millisecond-precision timestamps, `IntervalDayTime` type could be coerced, and for the other timestamp types `IntervalMonthDayNano` seems to be the only option.

**Pull request merged**